### PR TITLE
Feature/routing

### PR DESCRIPTION
**Pull request by Tsuyoshi-Ishikawa**

### CODE REVIEW

**Review by ironbeer — CHANGES_REQUESTED** [review]
Thanks for the pull request, I commented so please check it!

File: src/services/proxy.service.ts [hunk 1]
```diff
@@ -47,16 +49,22 @@ export class ProxyService {
...
-      if (method !== 'eth_sendRawTransaction') {
-        return await this.verseService.post(headers, body);
+      if (method !== 'eth_sendRawTransaction' && isUseReadNode) {
+        return await this.verseService.postVerseReadNode(headers, body);
+      } else if (method !== 'eth_sendRawTransaction' && !isUseReadNode) {
+        return await this.verseService.postVerseMasterNode(headers, body);
       }
 
-      return await this.sendTransaction(headers, body);
+      return await this.sendTransaction(isUseReadNode, headers, body);
```
[the remark this answers is not in the document]
Comment (Tsuyoshi-Ishikawa) [reply src/services/proxy.service.ts:67]: If I don't use isUseReadNode in this.sendTransaction, it always goes to masterNode when doing estimatesGas, is that OK?

https://github.com/oasysgames/verse-proxy/blob/c0ad05fd5584cfbf72ee88fb42f3a08c3d670be2/src/services/transaction.service.ts#L124-L129

File: src/controllers/proxy.controller.ts [hunk 2]
```diff
@@ -28,10 +30,53 @@ export class ProxyController {
...
+        headers,
+        body,
+        callback,
+      );
+    } else {
+      throw new ForbiddenException(`invalid request`);
+    }
+  }
+
+  @Post('master')
```
[the remark this answers is not in the document]
Comment (Tsuyoshi-Ishikawa) [reply src/controllers/proxy.controller.ts:33]: Thanks.

I modified.
https://github.com/oasysgames/verse-proxy/pull/20/commits/eadb21595853abdd060f3a97ebd9fc8770e99c87

File: src/services/proxy.service.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (Tsuyoshi-Ishikawa) [reply src/services/proxy.service.ts]: Thanks!

I modified.
https://github.com/oasysgames/verse-proxy/pull/20/commits/165c4fc962a6462fc9dcd9e782d8f31fea5a6778

File: src/config/configuration.ts [hunk 3]
```diff
@@ -1,5 +1,6 @@
 export default () => ({
-  verseUrl: process.env.VERSE_URL ?? 'http://localhost:8545',
+  verseMasterNodeUrl: process.env.VERSE_MASTER_NODE_URL,
```
Comment (ironbeer) [inline src/config/configuration.ts:2]: Please add a fallback to the previous environment variable `VERSE_URL`

Comment (Tsuyoshi-Ishikawa) [reply src/config/configuration.ts]: Thanks!

I modified.
https://github.com/oasysgames/verse-proxy/pull/20/commits/9cf0db123549da7ff1c43b6cefb225fdab311f45

**Review by ironbeer — APPROVED** [review]
👍

**Pull request merged**